### DISCUSSION TITLE
Building a NLPModel with a full Hessian

### DISCUSSION THREAD

**Discussion by nrummel** [JuliaSmoothOptimizers/Organization#52] [General]
Hello JSO Team, 

I am have an analytic form for the hessian, and I would like to build a NLPModel with my own in-place Julia functions rather than using automatic differentiation. In the [ManualNLPModels](https://github.com/JuliaSmoothOptimizers/ManualNLPModels.jl) repo it looks like you can do this by providing hessian-vector product (hvp) function. Because my hessian is expensive to compute, I would like to compute the Hessian once at a particular iteration then reuse it if multiply hvp's are necessary. I have a current work around 
```julia
H!(H,x) = ... # custom hessian fun
_x_mem = zeros(N)
_H_mem = zeros(N,N)
function hvp_with_mem!(h, x, v; obj_weight=1.0)
  if !all(x .== _x_mem)
    H!(_H_mem, x
    _x_mem .= x 
  end
  h .= _H_mem*v
  h 
end
```
This feels a bit hackey, but it definitely solve my problem. Do you all have more elegant way to handle this situation?
Nic

**Comment by dpo** [comment]
@nrummel you could implement `hess_structure!()` and `hess_coord!()`. Then your solver can request the explicit Hessian whenever it sees fit. If your Hessian is dense, this isn’t ideal either, but we are working on a better solution.

**Comment by tmigot** [comment]
Btw @nrummel you could use 
```
shah_mem = hash(_x_mem)
  if shah_mem != hash(x)
```
to avoid comparing both `x`, which can be costly if the dimension is large.
